Storage/src/Connection/RetryTrait.php

```diff
@@ -0,0 +1,244 @@
+        'default_object_acl.list',
+        'hmacKey.delete',
+        'hmacKey.get',
+        'hmacKey.list',
+        'notifications.delete',
+        'notifications.get',
+        'notifications.list',
+        'object_acl.get',
+        'object_acl.list',
+        'objects.get',
+        'objects.list',
+        'serviceaccount.get',
+    ];
+
+    /**
+     * The operations which can be retried with specific conditions
+     * (Conditionally idempotent)
+     * @var array
+     */
+    public static $condIdempotentOps = [
+        'buckets.patch' => ['ifMetagenerationMatch', 'etag'],
+        // Currently etag is not supported, so this preCondition never available
+        'buckets.setIamPolicy' => ['etag'],
+        'buckets.update' => ['ifMetagenerationMatch', 'etag'],
+        'hmacKey.update' => ['etag'],
+        'objects.compose' => ['ifGenerationMatch'],
+        'objects.copy' => ['ifGenerationMatch'],
+        'objects.delete' => ['ifGenerationMatch'],
+        'objects.insert' => ['ifGenerationMatch', 'ifGenerationNotMatch'],
+        'objects.patch' => ['ifMetagenerationMatch', 'etag'],
+        'objects.rewrite' => ['ifGenerationMatch'],
+        'objects.update' => ['ifMetagenerationMatch']
+    ];
+
+    /**
+     * Retry strategies which enforce certain behaviour like:
+     *     - Always retrying a call when an exception occurs(within the limits of 'max retries').
+     *     - Never retrying a call when an exception occurs.
+     *     - Retrying only when the operation is considered idempotent(default).
+     * These configurations are supplied for per api call basis.
+     *
+     * We can set $options['retryStrategy'] to one of "always", "never" and
+     * "idempotent". Anything apart from them is considered as "idempotent" and will be
+     * retried as intended.
+     */
+
+    /**
+     * Retry an API operation when an exception occurs if the exception has a retryable error code.
+     * @var string
+     */
+    public static $RETRY_STRATEGY_ALWAYS = 'always';
+
+    /**
+     * Never retry an API operation.
+     * @var string
+     */
+    public static $RETRY_STRATEGY_NEVER = 'never';
+
+    /**
+     * Retry an API operation only if it is considered idempotent
+     * and the exception has a retryable error code.
+     * @var string
+     */
+    public static $RETRY_STRATEGY_IDEMPOTENT = 'idempotent';
+
+    /**
+     * Return a retry decider function.
+     *
+     * @param string $resource resource name, eg: buckets.
+     * @param string $method method name, eg: get
+     * @param array $args
+     * @param callable $restRetryFunction User given retry function
+     * @return callable
+     */
+    public function getRestRetryFunction($resource, $method, array $args, $restRetryFunction = null)
+    {
+        if (isset($args['restRetryFunction'])) {
+            return $args['restRetryFunction'];
+        } elseif (!is_null($restRetryFunction)) {
+            return $restRetryFunction;
```

Comment:
Thank you for your suggestion. While I understand your point, the current implementation has been designed to allow for code reuse, as this method is called at multiple places with the same intent(rest > sendWithRetry, uploads and downloads). Therefore, extracting this logic and including in `getRestRetryFunction()` seemed like a more appropriate approach. However, I appreciate your feedback and let's discuss further if you have a better solution ☺️

Reply:
I feel quite strongly that this is not where this logic should be. Having 3 different defaults in a getter is not a good method API.

Looking at the code, the fourth argument `$restRetryFunction` is only supplied one time (in `sendWithRetry`). So this is not code reuse - it's only being used in one single place. It would be better to put that logic in the call for `sendWithRetry` like this:

```php
        if (!isset($args['restRetryFunction'])) {
            $args['restRetryFunction'] = $this->restRetryFunction ?: $this->getRestRetryFunction(
                $retryResource,
                $method,
                $args
            );
        }
```